### DISCUSSION TITLE
How to install SSH2 with Homebrew

### DISCUSSION THREAD

**Discussion by kiusau** [Homebrew/discussions#2829] [Everyday usage]
Hi, Community!

If I have understood correctly, one should use the following command to install PHP extensions when working in a Homebrew environment:  pecl install [package name].  What does one do when the PECL package compatible with PHP8 (beta version) is not available through the pecl application? The packaged extension that I would like to install is ssh2-1.3.1 found at https://pecl.php.net/package/ssh2

Under the assumption that my Homebrew installed PECL application will one day allow me to use the PECL command to install a stable version of SSH2 that is compatible with PHP8, what would you'all recommend for the interim?

Roddy

**Comment by SMillerDev** [comment]
You can install ssh2 with pecl just fine. Alternatively, you might want to install it from https://github.com/shivammathur/homebrew-extensions

**Reply by kiusau** [reply]
Thank you for responding, but the installation depends on the version of PHP that you are running.
```
 ~ % pecl install ssh2
pecl/ssh2 requires PHP (version >= 4.0.0, version <= 6.0.0), installed version is 8.1.1
No valid packages found
install failed
 ~ %
```
And, when I tried installing with ssh2-beta the installation failed
```
 % pecl install ssh2-beta

Build process completed successfully
Installing '/usr/local/Cellar/php/8.1.1/pecl/20210902/ssh2.so'

Warning: mkdir(): File exists in System.php on line 294
PHP Warning:  mkdir(): File exists in /usr/local/Cellar/php/8.1.1/share/php/pear/System.php on line 294

Warning: mkdir(): File exists in /usr/local/Cellar/php/8.1.1/share/php/pear/System.php on line 294
ERROR: failed to mkdir /usr/local/Cellar/php/8.1.1/pecl/20210902
 ~ %
```
Roddy

**Reply by SMillerDev** [reply]
```
Build process completed successfully
Installing '/usr/local/Cellar/php/8.1.1/pecl/20210902/ssh2.so'

Warning: mkdir(): File exists in System.php on line 294
PHP Warning: mkdir(): File exists in /usr/local/Cellar/php/8.1.1/share/php/pear/System.php on line 294

Warning: mkdir(): File exists in /usr/local/Cellar/php/8.1.1/share/php/pear/System.php on line 294
ERROR: failed to mkdir /usr/local/Cellar/php/8.1.1/pecl/20210902
```

It completed succesfully according to the message. Try checking if the file is actually installed.

**Comment by kiusau** [comment]
At the address /usr/local/Cellar/php/8.1.1 there is a symlink (alias)

lrwxr-xr-x   1 username  staff     23 Dec 26 13:09 pecl -> /usr/local/lib/php/pecl

And at the address to which it points -- namely, usr/local/lib/php/

there are two more symlink (alias) pointing back

lrwxr-xr-x  1 username  admin  39 Dec 26 13:09 20210902 -> ../../Cellar/php/8.1.1/lib/php/20210902
lrwxr-xr-x  1 username  admin  36 Dec 26 13:09 build -> ../../Cellar/php/8.1.1/lib/php/build

Neither of these yield anything resembling the desired PHP extension.

**Reply by kiusau** [reply]
What if I were to create the folder /usr/local/Cellar/php/8.1.1/pecl manually and run the procedure again?

**Reply by kiusau** [reply]
I fear that Homebrew would object.

**Reply by gromgit** [reply]
What's the output of `ls -ld /usr/local/lib/php/pecl`? If it doesn't exist, or is anything other than a directory, your `brew install php` post-install phase did not complete for some reason, in which case the proper fix is:
```
rm -f /usr/local/lib/php/pecl   # if it exists but isn't a directory
brew postinstall php
```

**Reply by kiusau** [reply]
Hi gromgit!

Before I take up your suggestion I would like to make note of the following:

~ % brew doctor
Your system is ready to brew.
~ %

**Reply by gromgit** [reply]
That's good, but in case there's any confusion, I'm saying that your `php` installation is half-b0rked, not Homebrew itself.

**Reply by kiusau** [reply]
OK.  So, I tried your suggestion.

% brew postinstall php
==> Postinstalling php
Warning: The post-install step did not complete successfully
You can try again using:
  brew postinstall php
 % 

Got any more ideas?

**Reply by SMillerDev** [reply]
That's clearly the problem then. Try running it with `--debug`

**Comment by kiusau** [comment]
After a little sleep.
```shell
~ % brew postinstall php --debug
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/php.rb
==> Postinstalling php
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FromPathLoader): loading /usr/local/opt/php/.brew/php.rb
Warning: The post-install step did not complete successfully
You can try again using:
  brew postinstall php
==> An exception occurred within a child process:
  Errno::EEXIST: File exists @ syserr_fail2_in - /usr/local/Cellar/php/8.1.1/pecl

/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:363:in `symlink'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:363:in `block in ln_s'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:1574:in `fu_each_src_dest0'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:361:in `ln_s'
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/php.rb:262:in `post_install'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1131:in `block (2 levels) in run_post_install'
/usr/local/Homebrew/Library/Homebrew/formula.rb:948:in `with_logging'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1130:in `block in run_post_install'
/usr/local/Homebrew/Library/Homebrew/utils.rb:588:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1119:in `run_post_install'
/usr/local/Homebrew/Library/Homebrew/postinstall.rb:22:in `<main>'
iwato@jupiter ~ % 
```

**Reply by gromgit** [reply]
`rm -f /usr/local/Cellar/php/8.1.1/pecl`, then run the postinstall again.

**Reply by kiusau** [reply]
~ % rm -f /usr/local/Cellar/php/8.1.1/pecl
~ % brew postinstall php
==> Postinstalling php
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set php_ini /usr/local/etc/php/8.1/php.ini system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set php_dir /usr/local/share/pear system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set doc_dir /usr/local/share/pear/doc system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set ext_dir /usr/local/lib/php/pecl/20210902 system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set bin_dir /usr/local/opt/php/bin system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set data_dir /usr/local/share/pear/data system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set cfg_dir /usr/local/share/pear/cfg system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set www_dir /usr/local/share/pear/htdocs system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set man_dir /usr/local/share/man system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set test_dir /usr/local/share/pear/test system
==> /usr/local/Cellar/php/8.1.1/bin/pear config-set php_bin /usr/local/opt/php/bin/php system
==> /usr/local/Cellar/php/8.1.1/bin/pear update-channels
~ %

**Reply by kiusau** [reply]
Shall I try pecl install ssh2-1.3.1 or pecl install ssh2-beta again?

**Reply by kiusau** [reply]
What exactly does the postinstall command achieve.  It looks like it found a problem in php_ini file and corrected it. Is this what a postinstall does:  look for problems in an installation and tries to correct them?

**Reply by SMillerDev** [reply]
`postinstall` does install steps that _need_ to be done on the user machine. In this case it's setting the configurations to the correct Homebrew values because that can't be done in the `install` definition.

Now that the configuration is correct you can try `pecl install ssh2-beta` again.

**Reply by kiusau** [reply]
~ % pecl install ssh2-beta
downloading ssh2-1.3.1.tgz ...
Starting to download ssh2-1.3.1.tgz (33,838 bytes)
.........done: 33,838 bytes
6 source files, building
running: phpize
.
.
.
Build process completed successfully
Installing '/usr/local/Cellar/php/8.1.1/pecl/20210902/ssh2.so'
install ok: channel://pecl.php.net/ssh2-1.3.1
Extension ssh2 enabled in php.ini
~ %

**Reply by kiusau** [reply]
What a nice way to wake up in the morning!
Thank you!

Roddy

**Reply by kiusau** [reply]
... 20210902 % ls -l
total 224
-rw-r--r--  1 username  admin  111504 Jan 21 08:31 ssh2.so
... 20210902 %

**Reply by kiusau** [reply]
Thank you, all!
